tests/basic.py

```diff
@@ -0,0 +1,133 @@
+    p.expect(f"/dev/{device.hidraw}: Nitrokey 3 {device.serial}")
+    # TODO: assert that there are no other keys
+
+
+class TestFido2(ExecUpgradeTest):
+    # TODO:
+    # - Test server with non-registered client
+    # - Test client with non-registered server
+    # - Test with multiple credentials
+
+    @contextmanager
+    def context(self, device):
+        yield Fido2(device.hidraw)
+
+    def prepare(self, fido2):
+        return fido2.register(b"user_id", "A. User")
+
+    def verify(self, fido2, credential):
+        fido2.authenticate([credential])
+
+
+class TestSsh(UpgradeTest):
+    def __init__(self, type: str):
+        self.type = type
+
+    @contextmanager
+    def context(self, device):
+        with TemporaryDirectory() as d:
+            yield (device, d)
+
+    def prepare(self, context):
+        (device, d) = context
+        return keygen(d, self.type)
+
+    def verify(self, context, state):
+        (device, d) = context
+        (key, pubkey) = state
+        (key_path, pubkey_path) = keypair(d, key, pubkey)
+        with authorized_key(pubkey):
+            p = spawn(ssh_command(pubkey_path, "whoami"))
+            p.expect(SSH_USER)
+
+
+@pytest.mark.parametrize("type", SSH_KEY_TYPES)
+def test_ssh(device, type) -> None:
+    TestSsh(type).run(device)
+
+
+class TestSshResident(UpgradeTest):
+    def __init__(self, type: str):
+        # TODO: PIN generation
+        self.type = type + "-sk"
+        self.pin = "".join(random.choices(string.digits, k=8))
+
+    @contextmanager
+    def context(self, device):
+        with TemporaryDirectory() as d:
+            yield (device, d)
+
+    def prepare(self, context):
+        (device, d) = context
+        p = spawn("nitropy fido2 set-pin")
+        p.expect("enter new pin")
+        p.sendline(self.pin)
+        p.expect("confirm new pin")
+        p.sendline(self.pin)
+        p.expect("done")
+        return keygen(d, self.type, resident=True, pin=self.pin)
+
+    def verify(self, context, state):
+        (device, d) = context
+        (key, pubkey) = state
+        (key_path, pubkey_path) = keypair(d, key, pubkey)
+        with authorized_key(pubkey):
+            p = spawn(ssh_command(pubkey_path, "whoami"))
+            p.expect(SSH_USER)
+
+        filename = "id_" + self.type.replace('-', '_') + "_rk"
+        download_dir = os.path.join(d, "download")
+        os.mkdir(download_dir)
+        pwd = os.getcwd()
+        try:
+            os.chdir(download_dir)
+            p = spawn("ssh-keygen -K")
+            p.expect("Enter PIN for authenticator")
+            p.sendline(self.pin)
+            p.expect("Enter passphrase")
+            p.sendline("")
+            p.expect("Enter same passphrase")
+            p.sendline("")
+            p.expect(filename)
+            assert os.path.exists(filename)
+            # TODO: check why the key is partially different
+            # with open(filename, "rb") as f:
+            #     assert f.read() == key
```

Comment:
@daringer @szszszsz Do you know why the private key generated by `ssh-keygen` and the private key returned by `ssh-keygen -K` are slightly different?  Is there are way we can compare them properly?

```
   b'-----BEGIN OPENSSH PRIVATE KEY-----\nb3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAE'                                                                                                                  
   b'bm9uZQAAAAAAAAABAAAASgAAABpzay1zc2\ngtZWQyNTUxOUBvcGVuc3NoLmNvbQAAACBxvYS'        
-  b'WmP8qVHYvYrrtq0G2dk3cs2C0NCWQAdsh\nnSxx7gAAAARzc2g6AAAA6DYbg9o2G4PaAAAAGn'                                                                                                                  
?                                                            ^^^^^^^^^^^                                                                                                                         
+  b'WmP8qVHYvYrrtq0G2dk3cs2C0NCWQAdsh\nnSxx7gAAAARzc2g6AAAA6CKONykijjcpAAAAGn'        
?                                                            ^^^^^^^^^^^                                                                                                                         
   b'NrLXNzaC1lZDI1NTE5QG9wZW5zc2guY2\n9tAAAAIHG9hJaY/ypUdi9iuu2rQbZ2TdyzYLQ0J'                                                                                                                  
   b'ZAB2yGdLHHuAAAABHNzaDohAAAAhKMA\nWGB5dW9S6P9NISJxwyQ+Y/7wPS7v2PATs7V/PWRu'                                                                                                                  
   b'VzzGAXsvKhkRc0ir+jMI/OA+q+aLbX\nzf8zQFgpowo4K3M3DVvmte6GCSNU1Gei6G4iQg3qz'                                                                                                                  
   b'AVfyOvwGUTyPRjFS+I4QBTNs+mGFK\n+H04ITk6WgJQG5nDDQl4/Gf+LCnhmBSyRgAAAAAAAA'                                                                                                                  
-  b'AEZmlkbwE=\n-----END OPENSSH PRIVATE KEY-----\n',                                                                                                                                           
?      ^^^^^^                                                                                                                                                                                    
+  b'AEc3NoOgE=\n-----END OPENSSH PRIVATE KEY-----\n',
?
```